File: app/src/main/java/com/mobarak/todo/ui/taskdetail/TaskDetailViewModel.java
package com.mobarak.todo.ui.taskdetail;

import android.content.Context;
import android.util.Log;
import android.view.View;

import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.ViewModel;
import androidx.navigation.NavDirections;
import androidx.navigation.Navigation;

import com.mobarak.todo.R;
import com.mobarak.todo.data.AppRepository;
import com.mobarak.todo.data.db.entity.Task;
import com.mobarak.todo.ui.base.BaseViewModel;
import com.mobarak.todo.ui.tasks.FilterType;
import com.mobarak.todo.ui.tasks.TasksFragmentDirections;

import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.schedulers.Schedulers;

public class TaskDetailViewModel extends BaseViewModel {

    private static final String TAG = TaskDetailViewModel.class.getSimpleName();

    public TaskDetailViewModel(Context context, AppRepository repository) {
        super(context, repository);
    }

    private MutableLiveData<Long> _taskId = new MutableLiveData<Long>();

    public MutableLiveData<Task> task = new MutableLiveData<>();

    public MutableLiveData<Boolean> isDataAvailable = new MutableLiveData<>(false);
    public MutableLiveData<Boolean> dataLoading = new MutableLiveData<>(false);
    public MutableLiveData<Boolean> completed = new MutableLiveData<>(false);
    public MutableLiveData<String> editTaskEvent = new MutableLiveData<>();
    public MutableLiveData<String> deleteTaskEvent = new MutableLiveData<>();
    public MutableLiveData<String> snackbarText = new MutableLiveData<>();


    public void deleteTask() {
        if (_taskId.getValue() != null)
            mDisposable.add(repository.getDbRepository().deleteTaskById(_taskId.getValue())
                    .subscribeOn(Schedulers.io())
                    .observeOn(AndroidSchedulers.mainThread())
                    .subscribe(() -> {
                                snackbarText.setValue(context.getString(R.string.menu_delete_task));
                            },
                            throwable -> {
                                Log.e(TAG, "Task deleted", throwable);
                            }));
    }

    public void editTask(View view) {
        if (_taskId.getValue() != null) {
            NavDirections action = TaskDetailFragmentDirections
                    .actionTaskDetailFragmentToAddEditTaskFragment(_taskId.getValue(), context.getString(R.string.add_task));
            Navigation.findNavController(view).navigate(action);
        }
    }

    public void setCompleted(Boolean completed) {
        Task t = task.getValue();
        if (completed) {
            //repository.completeTask(task);
            showSnackbarMessage(context.getString(R.string.task_marked_complete));
        } else {
            //tasksRepository.activateTask(task)
            showSnackbarMessage(context.getString(R.string.task_marked_active));
        }
    }

    public void start(long taskId) {
        // If we're already loading or already loaded, return (might be a config change)
        if ((dataLoading.getValue() != null && dataLoading.getValue())
                || (_taskId.getValue() != null && taskId == _taskId.getValue())) {
            return;
        }
        // Trigger the load
        _taskId.setValue(taskId);
        loadTask();
    }

    private void loadTask() {
        mDisposable.add(repository.getDbRepository().observeTaskById(_taskId.getValue())
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(item -> {
                            task.setValue(item);
                            isDataAvailable.setValue(true);
                        },
                        throwable -> {
                            Log.e(TAG, "no task found", throwable);
                        }));
    }

    private Task computeResult() {
        return null;
    }

    public void refresh() {
        // Refresh the repository and the task will be updated automatically.
        if (task != null && task.getValue() != null) {
        }
    }

    private void showSnackbarMessage(String message) {
        snackbarText.setValue(message);
    }
}

File: app/src/main/java/com/mobarak/todo/ui/taskdetail/TaskDetailViewModel.kt
package com.mobarak.todo.ui.taskdetail

import androidx.annotation.StringRes
import androidx.lifecycle.*
import com.mobarak.todo.R
import com.mobarak.todo.data.AppRepository
import com.mobarak.todo.data.db.entity.Task
import com.mobarak.todo.utility.Event
import kotlinx.coroutines.launch

class TaskDetailViewModel(private val repository: AppRepository) : ViewModel() {
    private val _taskId = MutableLiveData<Long>()

    private val _task = _taskId.switchMap { taskId ->
        repository.getDbRepository().observeTaskById(taskId).asLiveData().map {
            computeResult(it)
        }
    }
    val task: LiveData<Task?> = _task

    val isDataAvailable: LiveData<Boolean> = _task.map { it != null }

    private val _dataLoading = MutableLiveData<Boolean>()
    val dataLoading: LiveData<Boolean> = _dataLoading

    private val _editTaskEvent = MutableLiveData<Event<Unit>>()
    val editTaskEvent: LiveData<Event<Unit>> = _editTaskEvent

    private val _deleteTaskEvent = MutableLiveData<Event<Unit>>()
    val deleteTaskEvent: LiveData<Event<Unit>> = _deleteTaskEvent

    private val _snackbarText = MutableLiveData<Event<Int>>()
    val snackbarText: LiveData<Event<Int>> = _snackbarText

    // This LiveData depends on another so we can use a transformation.
    val completed: LiveData<Boolean> = _task.map { input: Task? ->
        input?.isCompleted ?: false
    }

    fun deleteTask() = viewModelScope.launch {
        _taskId.value?.let {
            repository.getDbRepository().deleteTaskById(it)
            _deleteTaskEvent.value = Event(Unit)
        }
    }

    fun editTask() {
        _editTaskEvent.value = Event(Unit)
    }

    fun setCompleted(completed: Boolean) = viewModelScope.launch {
        val task = _task.value ?: return@launch
        if (completed) {
            repository.getDbRepository().updateTask(task)
            showSnackbarMessage(R.string.task_marked_complete)
        } else {
            repository.getDbRepository().updateTask(task)
            showSnackbarMessage(R.string.task_marked_active)
        }
    }

    fun start(taskId: Long) {
        // If we're already loading or already loaded, return (might be a config change)
        if (_dataLoading.value == true || taskId == _taskId.value) {
            return
        }
        // Trigger the load
        _taskId.value = taskId
    }

    private fun computeResult(task: Task): Task? {
        return if (task != null) {
            task
        } else {
            showSnackbarMessage(R.string.loading_tasks_error)
            null
        }
    }

    fun refresh() {
        // Refresh the repository and the task will be updated automatically.
        _task.value?.let {
            _dataLoading.value = true
            viewModelScope.launch {
                repository.getDbRepository().getTaskById(it.id)
                _dataLoading.value = false
            }
        }
    }

    private fun showSnackbarMessage(@StringRes message: Int) {
        _snackbarText.value = Event(message)
    }

    companion object {
        private val TAG = TaskDetailViewModel::class.java.simpleName
    }
}